core/src/providers/openai.rs

```diff
@@ -1015,79 +1222,554 @@ pub async fn chat_completion(
-                        sleep: Duration::from_millis(500),
-                        factor: 2,
+                                    // If we get `content` in the delta object we stream "tokens".
+                                    match chunk.choices[0].delta.get("content") {
+                                        None => (),
+                                        Some(content) => match content.as_str() {
+                                            None => (),
+                                            Some(s) => {
+                                                if s.len() > 0 {
+                                                    let _ = sender.send(json!({
+                                                        "type": "tokens",
+                                                        "content": {
+                                                            "text": s,
+                                                        },
+                                                    }));
+                                                }
+                                            }
+                                        },
+                                    };
+
+                                    // If we have exactly one `tool_call.function.name` in the delta object, stream a "function_call" event.
+                                    if let Some(tool_calls) = chunk.choices[0]
+                                        .delta
+                                        .get("tool_calls")
+                                        .and_then(|v| v.as_array())
+                                    {
+                                        if tool_calls.len() == 1 {
+                                            if let Some(f) = tool_calls[0].get("function") {
+                                                if let Some(Value::String(name)) = f.get("name") {
+                                                    let _ = sender.send(json!({
+                                                        "type": "function_call",
+                                                        "content": {
+                                                            "name": name,
+                                                        },
+                                                    }));
+                                                }
+                                            }
+                                        } else {
+                                            return Err(anyhow!("More than one tool call found"));
+                                        }
+                                    }
+                                }
+                            }
+                            None => (),
+                        };
+                        chunks.lock().push(chunk);
+                    }
+                },
+                None => {
+                    println!("UNEXPECED NONE");
```

Comment:
nit UNEXPECTED, but also shouldn't this be an actual error ?

Reply:
Copied/pasted from the previous implementation.